stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+            _Add_mutex_for_instance_or_increment(static_cast<void*>(_Wrapped));
+        }
+        _Init();
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
+        _Assign_rv(_STD move(_Right));
+    }
+
+    ~basic_syncbuf() {
+        _Emit();
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
+        _Emit();
+        _Move_assign(_STD move(_Right), _Choose_pocma<_Alloc>{});
+        return *this;
+    }
+
+    void swap(basic_syncbuf& _Right) {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
+            _Swap_except_al(_Right);
+        }
+    }
+
+    bool emit() {
+        if (_Wrapped == nullptr) {
+            return false;
+        }
+
+        bool _Result                 = true;
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(static_cast<void*>(_Wrapped)));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {
+                _Result = false;
+            }
+
+            if (_Record_sync) {
+                if (_Wrapped->pubsync() == -1) {
+                    _Result = false;
+                }
+            }
+        }
+        _Record_sync = false;
+        streambuf_type::setp(
+            _Begin_seq_ptr, pointer_traits<pointer>::pointer_to(*streambuf_type::epptr())); // reset written data
+        return _Result;
+    }
+
+    streambuf_type* get_wrapped() const noexcept {
+        return _Wrapped;
+    }
+
+    allocator_type get_allocator() const noexcept {
+        return _Getal();
+    }
+
+    void set_emit_on_sync(bool _Val) noexcept {
+        _Emit_on_sync = _Val;
+    }
+
+protected:
+    int sync() override {
+        _Record_sync = true;
+
+        if (_Emit_on_sync) {
+            if (!emit()) {
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    int_type overflow(int_type _Current_elem) override {
+        if (_Wrapped == nullptr)
+            return _Traits::eof();
+        const bool _Chk_eof = _Traits::eq_int_type(_Current_elem, _Traits::eof());
+        if (_Chk_eof)
+            return _Traits::not_eof(_Current_elem);
+
+        const size_type _Buf_size       = _Get_buffer_size();
+        const size_type _Max_allocation = allocator_traits<_Alloc>::max_size(_Al);
+        if (_Buf_size == _Max_allocation)
+            return _Traits::eof();
+
+        const size_type _New_capacity  = _Calculate_growth(_Buf_size, _Buf_size + 1, _Max_allocation);
+        const pointer _Old_ptr         = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        const size_type _Old_data_size = _Get_data_size();
+
+        const pointer _New_ptr = _Unfancy(_Al.allocate(_New_capacity));

```

Comment:
Is this correct? `_Unfancy` returns a raw pointer, but `pointer` is the allocator's fancy pointer type. One of these seems incorrect (either we need `pointer` so we shouldn't unfancy, or we need to unfancy so we should be storing a raw pointer). Occurs below. We probably need test coverage for fancy pointers here.

Reply:
Yep, I think this is incorrect since I am using the same `pointer` type from the allocator.